onnxruntime/core/framework/callback.h

```diff
@@ -33,4 +34,37 @@ struct OrtCallbackInvoker {
     }
   }
 };
-}  // namespace onnxruntime
\ No newline at end of file
+
+/**
+ * Invokes the contained OrtCallback upon destruction or being assigned to.
+ */
+class ScopedOrtCallbackInvoker {
+ public:
+  explicit ScopedOrtCallbackInvoker(OrtCallback callback)
+      : callback_(callback) {}
+
+  ScopedOrtCallbackInvoker(ScopedOrtCallbackInvoker&& other)
+      : callback_(other.callback_) {
+    other.callback_.f = nullptr;
+    other.callback_.param = nullptr;
+  }
+
+  ScopedOrtCallbackInvoker& operator=(ScopedOrtCallbackInvoker&& other) {
+    if (callback_.f) callback_.f(callback_.param);
+
+    callback_ = other.callback_;
+    other.callback_.f = nullptr;
+    other.callback_.param = nullptr;
+
+    return *this;
+  }
+
+  ~ScopedOrtCallbackInvoker() {
+    if (callback_.f) callback_.f(callback_.param);
+  }
```

Comment:
nit: put the callback_.f(...) on the next line so a breakpoint can be set on that. same for line 53.